### PR TITLE
[DM-28119] Gafaelfawr cleanups

### PR DESCRIPTION
**Pull request by rra**
* Document new parameters returned by login route
* Document GAFAELFAWR_UI_PATH
* Use react-alert for UI alerts
* Add a matrix for database testing and test against PostgreSQL in GitHub Actions
* Parse API errors and associate them with form fields
* Reissue half-expired tokens during token delegation
* Shorten expiration of subtokens of user tokens whose expiration has decreased
* Add title and description to all model fields and routes
* Expose the API documentation of the running service
* Investigate fixing token deletion to check whether the token was in Redis
* Use more specific bad request status codes
* Remove logging that's now pointless given uvicorn logging
* Don't require CSRF tokens when using bearer token authentication
* Cascading delete of tokens
* Support URL query parameters for token history page
* Handle invalid group and scope names
* Refactor error handling for bad token creation to avoid duplicate code
* Add a scope for user token manipulation
* Check that unset {{auth_request}} headers are stripped by NGINX
* The modify token event logs raw datetimes instead of timestamps
* Impose username requirements of alphanumerics, single hyphen, cannot start or end with hyphen (same as GitHub)
* Use nvm to install Node in the Docker image

### CODE REVIEW

**Comment by rra** [comment]
This is a bunch of fairly separate fixes, so it may be easier to review commit by commit.

**Review by athornton — APPROVED** [review]
LGTM

**Review by athornton — APPROVED** [review]
Seems legit.

**Review by athornton — APPROVED** [review]
LGTM

**Review by athornton — APPROVED** [review]
LGTM

**Review by athornton — APPROVED** [review]
I didn't follow all of that, but it got a lot shorter, which I like.  To the extent I got it, it looks fine.

**Review by athornton — APPROVED** [review]
LGTM

**Review by athornton — APPROVED** [review]
Very lucid explanation in the docs!

File: src/gafaelfawr/constants.py [hunk 1]
```diff
@@ -15,14 +15,20 @@
 SETTINGS_PATH = "/etc/gafaelfawr/gafaelfawr.yaml"
 """Default configuration path."""
 
-ACTOR_REGEX = "^(?:<bootstrap>|[a-z0-9._-]+)$"
+# The following constants are used for field validation.  Minimum and maximum
+# length are handled separately.
+
+ACTOR_REGEX = "^(?:<bootstrap>|[a-z_][a-z0-9._-]+)$"
 """"Regex matching all valid actors (including ``<bootstrap``)."""
```
Comment (athornton) [inline src/gafaelfawr/constants.py]: Close the angle bracket on bootstrap here?

**Review by athornton — APPROVED** [review]
One minor doc niggle, otherwise good.

**Review by athornton — APPROVED** [review]
JavaScript not really my thing, but this seems straightforward.

**Review by athornton — APPROVED** [review]
This seems pretty straightforward, although if you want someone actually fluent in JS/React.js, it's not me yet.

**Review by athornton — APPROVED** [review]
The allow-unsafe makes sense and I'm just going to assume that the hashes are all right.

**Review by athornton — APPROVED** [review]
LGTM, AFAICT

**Review by athornton — APPROVED** [review]
I actually think I understand the error-alerting simplification here.

**Review by athornton — APPROVED** [review]
LGTM

**Review by athornton — APPROVED** [review]
LGTM

**Review by athornton — APPROVED** [review]
LGTM

File: src/gafaelfawr/services/token.py [hunk 2]
```diff
@@ -817,12 +817,13 @@ async def modify_token(
                     )
 
         if info:
+            timestamp = int(info.expires.timestamp()) if info.expires else None
```
Comment (athornton) [inline src/gafaelfawr/services/token.py:822]: Very Perlish

**Review by athornton — APPROVED** [review]
LGTM

**Review by athornton — APPROVED** [review]
LGTM

**Review by athornton — APPROVED** [review]
LGTM

**Review by athornton — APPROVED** [review]
LGTM

**Review by athornton — APPROVED** [review]
My eyes kind of glazed over but LGTM

File: src/gafaelfawr/models/oidc.py [hunk 3]
```diff
@@ -156,3 +160,231 @@ class OIDCVerifiedToken(OIDCToken):
...
+            "c3Ygu_HWlfO56GppTE7A9fksYVMcaSdi6zVvWZH-PmgZAlPZ4xs4NQ_pXdIUA5yc4"
+            "NhLAoQ5jBkPCTXsr4tqTBkPKKXxsNDLUFeS262o58kvOAgSCDLRuFXVVHEUIOw-kG"
+            "ko_UGmG3O5R3o-dC7f7K3OOUI_UulCldWu1ZgdckUfIS7fyMDmcZ4vNL8EALDEewv"
+            "mjwyO_OLqquNFnfMeJdKNPw"
+        ),
+    )
+
+    id_token: str = Field(
+        ...,
+        title="Idnetity token",
```
Comment (athornton) [inline src/gafaelfawr/models/oidc.py:192]: typo

File: src/gafaelfawr/models/oidc.py [hunk 4]
```diff
@@ -156,3 +160,231 @@ class OIDCVerifiedToken(OIDCToken):
...
+            "M8hJHJcEkg9htWceQfTCPAG7kP9p0K_bF3BM-8zXw53eE7g3Nd8Yz3875PrPIG7Je"
+            "KWz7mef8YNmv331fXc"
+        ),
+    )
+
+    e: str = Field(
+        ...,
+        title="RSA exponent",
+        description=(
+            "Big-endian exponent componet of the RSA public key encoded in"
```
Comment (athornton) [inline src/gafaelfawr/models/oidc.py:299]: typo

**Review by athornton — APPROVED** [review]
Couple of single-char typos in the docs I'd like to see fixed but nothing that actually impedes function or understandability.

**Pull request merged**